src/formatDate.js

```diff
@@ -8,7 +8,76 @@
  * @returns {string}
  */
 function formatDate(date, fromFormat, toFormat) {
-  // write code here
+  const formatDictionary = {};
+  const oldSeparator = fromFormat[3];
+  const newSeparator = toFormat[3];
+  const inputDateParts = date.split(oldSeparator);
+  const newDateParts = [];
+
+  /* Create a dictionary with:
+   * - key - first letter of the part format ('D', 'M' or 'Y')
+   * - value - dictionary with the given 'format' and 'value' of this date part
+   */
+  for (let i = 0; i < 3; i++) {
+    const format = fromFormat[i];
+    const formatCategory = format[0];
+    const value = inputDateParts[i];
+
+    formatDictionary[formatCategory] = {
+      format: format,
+      value: value,
+    };
+  }
+
+  /* For each format part of the 'toFormat', use the first letter as the key
+   * to find the value that should be inserted into the same position and its
+   * initial format.
+   * If the value is the year number, check and if needed update the format.
+   */
+  for (let i = 0; i < 3; i++) {
+    const format = toFormat[i];
+    const formatCategory = format[0];
+    const oldFormat = formatDictionary[formatCategory]['format'];
+    let value = formatDictionary[formatCategory]['value'];
+
+    if (formatCategory === 'Y') {
+      value = convertYear(value, oldFormat, format);
+    }
+
+    newDateParts.push(value);
```

Comment:
The second loop to construct `newDateParts` is working fine, but it's tightly coupled with the structure of `formatDictionary`. If you simplify the `formatDictionary` as suggested earlier, you will need to adjust this loop accordingly.